x-pack/plugins/visualization_lens/public/types.ts

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+export interface EditorFrameAPI {
+  render: (domElement: Element) => void;
+  registerDatasource: (datasource: Datasource<unknown>) => void;
+  registerVisualization: (visualization: Visualization<unknown>) => void;
+}
+
+export interface EditorFrameState {
+  visualizations: { [id: string]: object };
+  datasources: { [id: string]: object };
+
+  activeDatasourceId: string;
+  activeVisualizationId: string;
+}
+
+// Hints the default nesting to the data source. 0 is the highest priority
+export type DimensionPriority = 0 | 1 | 2;
+
+// For switching between visualizations and correctly matching columns
+export type DimensionRole =
+  | 'splitChart'
+  | 'series'
+  | 'primary'
+  | 'secondary'
+  | 'color'
+  | 'size'
+  | string; // Some visualizations will use custom names that have other meaning
+
+export interface TableMetaInfo {
+  columnId: string;
+  operation: Operation;
+}
+
+export interface DatasourceSuggestion<T> {
+  state: T;
+  tableMetas: TableMetaInfo[];
+}
+
+/**
+ * Interface for the datasource registry
+ */
+export interface Datasource<T> {
+  // For initializing from saved object
+  // initialize: (state?: T) => Promise<T>;
+
+  renderDataPanel: (props: DatasourceDataPanelProps) => void;
+
+  toExpression: (state: T) => string;
+
+  getDatasourceSuggestionsForField: (state: T) => Array<DatasourceSuggestion<T>>;
+  getDatasourceSuggestionsFromCurrentState: (state: T) => Array<DatasourceSuggestion<T>>;
+
+  getPublicAPI: (state: T, setState: (newState: T) => void) => DatasourcePublicAPI;
+  // publicAPI: {
+  //   [key in keyof DatasourcePublicAPI]: (
+  //     state: T,
+  //     setState: (newState: T) => void
+  //   ) => DatasourcePublicAPI[key]
+  // };
+}
+
+/**
+ * This is an API provided to visualizations by the frame, which calls the publicAPI on the datasource
+ */
+export interface DatasourcePublicAPI {
+  getTableSpec: () => TableSpec;
+  getOperationForColumnId: () => Operation;
+
+  // Called by dimension
+  getDimensionPanelComponent: (
+    props: DatasourceDimensionPanelProps
+  ) => (domElement: Element, operations: Operation[]) => void;
```

Comment:
It feels like operations should be part of props, not an argument to the internal function.

Reply:
I think this entire definition is backwards. Shouldn't it be:

```
getDimensionPanelComponent: (domElement: Element) => (props) => void;
```

Reply:
Shouldn’t it be just a single function (props, domElement) => void;